### DISCUSSION TITLE
Bug or feature? Linebreaking fails for specific char combinations

### DISCUSSION THREAD

**Discussion by ghost** [typst/typst#5485] [Q&A] [answered]
I currently use typst to create a full font table of all codepoint pairs (c1, c2) to have a quick overview where I should change the default metric of c1 to minimize the required kerning pairs. For example the letter A gets A1 A2 A3 ... until all codepoints are combined with A. This is repeated for all other chars. For all pairs there is a white space appended to allow linebreaking. If c1 is finished an additional explicit linebreak is done.

Problem:
This works for all combinations as expected but for some reason it fails the following to linebreak:
)X
,X
.X
:X
;X
]X

where X is some other char. Even though there is always a white space between this two-char-pair, which makes it quite simple to line break.

Is this a bug or a feature? For me it seems that these chars are chosen explicitly since they are some kind of delimiters. Can I workaround this to also allow linebreak for those?

I think the following code can reproduce it a bit smaller:

```typst
#set page(width: 2cm)
#set par(justify: true)

#let chars = "ABCDEF),.:;]X"
#{

    for c1 in chars {
        for c2 in chars {
            c1 + c2 + [ ]
        }
        linebreak()
    }

}
```

We can see that the letters work as expected and the symbols fail. This is not font dependend.
![fail](https://github.com/user-attachments/assets/96040bca-2594-43cd-9fd7-7b926b456c2d)

**Comment by Enivex** [comment, marked as answer]
There are rules for where linebreaks are allowed to occur in text. Typically you don't want new lines starting with a comma for instance. I'm pretty sure that's what's going on here.

When writing something that's not actually text I'd go for a grid/table instead of relying on regular text layout.

So I'm going with feature, not a bug.

Wrapping each pair in a box might also work, though I haven't tried.

**Reply by ghost** [reply]
Makes much sense, thanks!

I found a fix by adding a `sym.zws` in front:

```typst
#set page(width: 2cm)
#set par(justify: false)

#let chars = "ABCDEF),.:;]X"
#{

    for c1 in chars {
        for c2 in chars {
            sym.zws + c1 + c2 + [ ]
        }
        linebreak()
    }

}
```
